From 16bfb9b92099e154cc2c5fdb858adb13107cb993 Mon Sep 17 00:00:00 2001
From: Yongli He <yongli.he@windriver.com>
Date: Wed, 6 Jul 2011 15:19:00 +0800
Subject: [PATCH 3/8] fsl_p50x0 fix call trace on preempt kernel

checking missing dbell_exception need disable preempt.

BUG: using smp_processor_id() in preemptible [00000000] code: bash/707
caller is .doorbell_exception+0x94/0x1c0
Call Trace:
[c0000000f91476c0] [c000000000009880] .show_stack+0x60/0x1a0 (unreliable)
[c0000000f9147760] [c000000000435fe4] .debug_smp_processor_id+0x104/0x120
[c0000000f9147800] [c000000000019ab4] .doorbell_exception+0x94/0x1c0
[c0000000f91478b0] [c000000000006498] .raw_local_irq_restore+0x48/0x90
[c0000000f9147930] [c00000000013c7e8] .kmem_cache_free+0xe8/0x190
[c0000000f91479d0] [c000000000125c88] .remove_vma+0x88/0xf0
[c0000000f9147a60] [c000000000125e98] .exit_mmap+0x1a8/0x230
[c0000000f9147b10] [c00000000004aff0] .mmput+0x70/0x160
[c0000000f9147ba0] [c000000000050d74] .exit_mm+0x194/0x1d0
[c0000000f9147c50] [c000000000052fb0] .do_exit+0x170/0x7f0
[c0000000f9147d30] [c000000000053680] .do_group_exit+0x50/0xd0
[c0000000f9147dc0] [c000000000053714] .SyS_exit_group+0x14/0x30
[c0000000f9147e30] [c0000000000005a4] syscall_exit+0x0/0x2c

Signed-off-by: Yongli He <yongli.he@windriver.com>
---
 arch/powerpc/kernel/irq.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/kernel/irq.c b/arch/powerpc/kernel/irq.c
index f81da1a..800a2d1 100644
--- a/arch/powerpc/kernel/irq.c
+++ b/arch/powerpc/kernel/irq.c
@@ -158,7 +158,9 @@ notrace void raw_local_irq_restore(unsigned long en)
 
 #if defined(CONFIG_BOOKE) && defined(CONFIG_SMP)
 	/* Check for pending doorbell interrupts on SMP */
+	preempt_disable();
 	doorbell_exception(NULL);
+	preempt_enable();
 #endif
 
 	/*
-- 
1.7.0.4

